### PR TITLE
Scalability part I: Multiports

### PR DESCRIPTION
**Pull request by edwardalee**
This branch does a major refactoring of the instance classes, ReactorInstance, ReactionInstance, PortInstance, etc., converting them all to Java, eliminating the MultiportInstance class, and consolidating so that a multiport results in exactly one PortInstance being created rather than one PortInstance per channel.  This is a large PR with a lot of changes, but it is still a work in progress. Quite a bit more work needs to be done, most notably to handle banks more intelligently and create only one ReactorInstance per bank rather than one ReactorInstance per bank member.  That additional work will occur in a new branch derived from this one.

### CODE REVIEW

File: org.lflang/src/org/lflang/federated/FederateInstance.xtend
Comment (lhstrh) [inline org.lflang/src/org/lflang/federated/FederateInstance.xtend:358]: I'm confused about this comment. What code/repo is this comment referring to? The `reactor-ts` submodule only has TypeScript code in it...

File: org.lflang.diagram/src/org/lflang/diagram/synthesis/styles/LinguaFrancaShapeExtensions.xtend
Comment (lhstrh) [inline org.lflang.diagram/src/org/lflang/diagram/synthesis/styles/LinguaFrancaShapeExtensions.xtend:296]:
```suggestion
        if (SHOW_REACTION_LEVEL.booleanValue) {
```

File: org.lflang.diagram/src/org/lflang/diagram/synthesis/styles/LinguaFrancaShapeExtensions.xtend
Comment (lhstrh) [inline org.lflang.diagram/src/org/lflang/diagram/synthesis/styles/LinguaFrancaShapeExtensions.xtend:295]:
```suggestion
```

File: org.lflang/src/org/lflang/generator/ActionInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/ActionInstance.java:119]: Should the default be null?

File: org.lflang/src/org/lflang/generator/GeneratorBase.xtend
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/GeneratorBase.xtend:109]: Maybe this should be `_lf_GenDelay` to be consistent with other LF-related reserved keywords in C.

File: org.lflang/src/org/lflang/generator/GeneratorBase.xtend
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/GeneratorBase.xtend:1752]: Should we create an issue for this? I can take this up as an action item.

File: org.lflang/src/org/lflang/generator/ParameterInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/ParameterInstance.java:53]:
```suggestion
     * Create a parameter instance from the specified definition
```

File: org.lflang/src/org/lflang/generator/ParameterInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/ParameterInstance.java:69]: Would type cause an issue here (e.g., if the user defines `bank_index:string`)?

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/PortInstance.java:53]:
```suggestion
     * Create a port instance from the specified definition
```

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/PortInstance.java:63]:
```suggestion
     * Create a port instance from the specified definition
```

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/PortInstance.java:77]: This could be factored out into a private method.

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/PortInstance.java:99]:
```suggestion
                                "Width of multiport "+ getFullName() +" cannot be determined at compile time. Assuming 1."
```

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/PortInstance.java:168]: This method is too long.

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/PortInstance.java:96]: I think for the following comment to be accurate, this statement should perhaps be in a `try...catch`, with catch using the provided error reporter, if any:
> An error reporter, or null to throw exceptions.

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/PortInstance.java:115]:
```suggestion
     * Clear cached information about the connectivity of this port.
```

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/PortInstance.java:116]:
```suggestion
     * In particular, {@link #eventualDestinations()} and {@link #eventualSources()}
```

**Review by lhstrh — APPROVED** [review]
These changes improve the handling of multiports and port a number of classes to Java. These are both great improvements! That said, I believe that if we want to address the fundamental memory consumption problems we've been seeing, the approach capitalized on in this design -- to create a full instantiation tree in memory -- is not tenable. While working on the operational semantics, I've sketched out an alternative design that would use significantly less memory. Such an overhaul, however, would not be easy and obviously outside of the scope of this effort. Having a clean implementation of multiports and banks will certainly also be helpful as a reference for such alternative implementation, so I nonetheless think it's a good idea to merge this into master. I've only left a few minor comments.

**Comment by edwardalee** [comment]
I agree that creating a "full instantiation tree" is not tenable, and in fact, this PR is precisely addressing this issue by not instantiating all the channels of a multiport.  The scalability-banks branch (a work in progress) is further extending this by not instantiating all bank members.  Once these two are done, I do not believe further reduction will yield any benefit because the resulting graph structure is the minimum size required for diagramming, so the graph will need to be constructed anyway if we are to have diagramming.

File: org.lflang/src/org/lflang/generator/c/CGenerator.xtend
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/c/CGenerator.xtend:3904]: There seems to be duplicate function calls here that are also in `generateFederate`. I wonder if these two could be consolidated, or the latter could be moved to `federated.CGeneratorExtension.java`

**Review by Soroosh129 — APPROVED** [review]
This is indeed an impressive improvement over the existing design. Looking forward to seeing part II of this effort. I left a few comments and some suggestions. I hope they can be helpful.

File: org.lflang/src/org/lflang/generator/c/CGenerator.xtend
Comment (edwardalee) [reply org.lflang/src/org/lflang/generator/c/CGenerator.xtend:3904]: Yes, this is on the to-do list. I've added a FIXME.

File: org.lflang/src/org/lflang/federated/FederateInstance.xtend
Comment (edwardalee) [reply org.lflang/src/org/lflang/federated/FederateInstance.xtend:358]: This function has been renamed "contains" from "containsReaction", so this is here just for backward compability in another repo.  The rename has to be done in the other repo, and then everything aligned. I don't have the bandwidth to do this.  I've added a more detailed explanation.

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (edwardalee) [reply org.lflang/src/org/lflang/generator/PortInstance.java:168]: Noted.

File: org.lflang/src/org/lflang/generator/ActionInstance.java
Comment (edwardalee) [reply org.lflang/src/org/lflang/generator/ActionInstance.java:1]: Good idea. Noted but deferred.

File: org.lflang/src/org/lflang/generator/ActionInstance.java
Comment (edwardalee) [reply org.lflang/src/org/lflang/generator/ActionInstance.java:119]: This is set unconditionally in the constructor, so null is not the default.

File: org.lflang/src/org/lflang/generator/GeneratorBase.xtend
Comment (edwardalee) [reply org.lflang/src/org/lflang/generator/GeneratorBase.xtend:109]: Good idea. Done.

File: org.lflang/src/org/lflang/generator/GeneratorBase.xtend
Comment (edwardalee) [reply org.lflang/src/org/lflang/generator/GeneratorBase.xtend:1752]: It would be great if you would take this as an action item, but I suggest waiting until after scalability-banks is merged.

File: org.lflang/src/org/lflang/generator/PortInstance.java
Comment (edwardalee) [reply org.lflang/src/org/lflang/generator/PortInstance.java:96]: Thanks. Comment is obsolete. Removed.

File: org.lflang/src/org/lflang/federated/FederateInstance.xtend
Comment (lhstrh) [reply org.lflang/src/org/lflang/federated/FederateInstance.xtend:358]: I still don't understand. What repo are we talking about? Can you provide a link?

Comment (cmnrd) [reply org.lflang/src/org/lflang/federated/FederateInstance.xtend:358]: If the method is here only to preserve compatibility, it should be labeled with `@Deprecated`. However, I also don't see how the TS runtime uses this method.

File: org.lflang/src/org/lflang/generator/ReactorInstance.java
Comment (cmnrd) [inline org.lflang/src/org/lflang/generator/ReactorInstance.java:69]: This class is huge. A good rule of thumb seems to be to keep classes below 500 lines. Is there maybe a way to split this class up. Also several methods in this class are too long (The method on line 911 for instance). Generally, method bodies should fit on a single screen.

**Review by cmnrd — APPROVED** [review]
Looks good! This is quite difficult to review carefully due to the many changes made, but I didn't notice any read flags.

**Comment by cmnrd** [comment]
BTW: There is something fishy going with the history in this branch. The commits between 
2cbe84e and 631e261 look like they are duplicated from master. It looks as if they were rebased. Was this on purpose?

As for the effectiveness of the changes made, I did a quick performance test compiling the bank transaction benchmark with 1000 accounts. It looks like this branch improves execution time, but increases the memory footprint even further.
`time -v` reports 123.68s user time on `master` and 10.2GB memory usage. On `scalability`, it reports 62.45s user time and 12.2GB memory usage. I am not sure if the changes made in this branch are supposed to influence the size of generated C code, but it is 1.2GB large on both branches.

**Comment by edwardalee** [comment]
As for the fishy history, I have no idea... All I did was merge in master. Was this the wrong thing to do?

**Comment by edwardalee** [comment]
The interleaved connection in Banking foils the methods used in this PR, I think. It will require a bit more work to handle those efficiently.

**Comment by cmnrd** [comment]
> As for the fishy history, I have no idea... All I did was merge in master. Was this the wrong thing to do?

Merging master shouldn't be a problem. But it looks like what actually happened is not a merge, but a rebase (there is also no merge commit). I observed this happening a few time now (we also had this in one of Marten's branches), but I don't know what causes this. Consider for instance the commit ce5269 on master. It is the same as commit 631e261 on this branch, but they got different hashes. The two commits are duplicates. I don't know what is causing this though... How did you perform the merge? Directly with `git merge master`, or did you use some other tool?

**Comment by edwardalee** [comment]
I used `git merge master`, then resolved conflicts and committed them, then did either `git rebase --continue` or `git rebase --abort` (the latter apparently needed if the conflicts are resolved in favor of what's in master).

**Comment by cmnrd** [comment]
Oh, that explains it. When you are merging, you should use `git merge --continue` instead. I don't know what precisely happens if you start a merge and then continue with a rebase, but apparently it aborts the merge and rebases all the changes instead (i.e. duplicating the master commits and putting them on top of the branch).

**Comment by edwardalee** [comment]
Ugh, ok.  Does this need to be undone? If so, how?

**Comment by cmnrd** [comment]
No, I don't think that we need to fix it. Nothing actually broke, we will only have a few duplicated commits when we merge this into master. For fixing it we would need to rewrite the history of this branch, but his would come at the high risk of breaking something else.

**Comment by lhstrh** [comment]
I want to be extra clear here to avoid further misunderstandings. When you do a `git pull --rebase` (`--rebase` is implicit if you have `pull.rebase` set to `true` in your configuration) and conflicts arise, them after resolving conflicts, you do a `git rebase --continue`. If you do a `git merge`, then after resolving conflicts you do a `git merge --continue`. Git tells also you this, but I the `merge` vs. `rebase` difference in the suggested commands is easily missed.

I'm glad we were able to pinpoint this, because it likely also explains some other issues that you have been experiencing with Git that really puzzled me.

**Pull request merged**
